RealmSwift/SwiftUI.swift

```diff
@@ -0,0 +1,599 @@
+        }
+    }
+    /// The object to observe.
+    @ObservedObject private var storage: ObservableStorage<ObjectType>
+    /// A default value to avoid invalidated access.
+    private let defaultValue: ObjectType
+
+    /// :nodoc:
+    public var wrappedValue: ObjectType {
+        get {
+            if storage.value.realm == nil {
+                // if unmanaged return the unmanaged value
+                return storage.value
+            } else if storage.value.isInvalidated {
+                // if invalidated, return the default value
+                return defaultValue
+            }
+            // else return the frozen value. the frozen value
+            // will be consumed by SwiftUI, which requires
+            // the ability to cache and diff objects and collections
+            // during some timeframe. The ObjectType is frozen so that
+            // SwiftUI can cache state. other access points will thaw
+            // the ObjectType
+            return storage.value.freeze()
+        }
+        set {
+            storage.value = newValue
+        }
+    }
+    /// :nodoc:
+    public var projectedValue: Wrapper {
+        return Wrapper(wrappedValue: storage.value.isInvalidated ? defaultValue : storage.value)
+    }
+    /**
+     Initialize a RealmState struct for a given thread confined type.
+     - parameter wrappedValue The RealmSubscribable value to wrap and observe.
+     */
+    public init(wrappedValue: ObjectType) where ObjectType: ObjectKeyIdentifiable {
+        _storage = ObservedObject(wrappedValue: ObservableStorage(wrappedValue))
+        defaultValue = ObjectType()
+    }
+    /**
+     Initialize a RealmState struct for a given thread confined type.
+     - parameter wrappedValue The RealmSubscribable value to wrap and observe.
+     */
+    public init<V>(wrappedValue: ObjectType) where ObjectType == List<V> {
+        _storage = ObservedObject(wrappedValue: ObservableStorage(wrappedValue))
+        defaultValue = List()
+    }
+}
+
+@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
+extension Binding where Value: ObjectBase & ThreadConfined {
+    /// :nodoc:
+    public subscript<V>(dynamicMember member: ReferenceWritableKeyPath<Value, V>) -> Binding<V> where V: _ManagedPropertyType {
+        createBinding(wrappedValue, forKeyPath: member)
+    }
+}
+
+/// :nodoc:
+@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
+public protocol BoundCollection {
+    /// :nodoc:
+    associatedtype Value
+
+    /// :nodoc:
+    var wrappedValue: Value { get }
+}
+
+@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
+public extension BoundCollection where Value: RealmCollection {
+    /// :nodoc:
+    typealias Element = Value.Element
+    /// :nodoc:
+    typealias Index = Value.Index
+    /// :nodoc:
+    typealias Indices = Value.Indices
+
+    /// :nodoc:
+    func remove<V>(at index: Index) where Value == List<V> {
+        let list = self.wrappedValue.realm != nil ?
```

Comment:
This logic shows up in a lot of places and would be worth extracting to a function.